# Nhập môn khoa học dữ liệu đồ án cuối kỳ _ Nhóm 12

Thông tin nhóm:
- 22120367 - Lê Hồng Tiến
- 22120
- 22120
- 22120395 - Nguyễn Sinh Trực
- 22120399 - Nguyễn Lê Anh Tú


## Thu thập dữ liệu

## Khai báo các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Khám phá và tiền xử lý dữ liệu
### Đọc dữ liệu từ file vào dataframe.

In [2]:
file_name = 'arxiv_cs_papers_all.csv'
raw_df = pd.read_csv(file_name)
raw_df.head()

paper_id  \
0  http://arxiv.org/abs/2411.15100v2   
1  http://arxiv.org/abs/2411.18620v1   
2  http://arxiv.org/abs/2411.18616v1   
3  http://arxiv.org/abs/2411.07976v5   
4  http://arxiv.org/abs/2411.18615v1   

                                             authors              updated  \
0  Yixin Dong,Charlie F. Ruan,Yaxing Cai,Ruihang ...  2024-11-27 18:59:28   
1  Zhi Zhang,Srishti Yadav,Fengze Han,Ekaterina S...  2024-11-27 18:59:26   
2  Shengqu Cai,Eric Chan,Yunzhi Zhang,Leonidas Gu...  2024-11-27 18:58:52   
3  Mahmut S. Gokmen,Caner Ozcan,Moneera N. Haque,...  2024-11-27 18:58:41   
4  Zhi Zhang,Jiayi Shen,Congfeng Cao,Gaole Dai,Sh...  2024-11-27 18:58:22   

             published                                              title  \
0  2024-11-22 18:01:37  XGrammar: Flexible and Efficient Structured Ge...   
1  2024-11-27 18:59:26  Cross-modal Information Flow in Multimodal Lar...   
2  2024-11-27 18:58:52  Diffusion Self-Distillation for Zero-Shot Cust...   
3  2024-11-12 17:55:39  DINO-LG: A Task-Specific DINO Model for Corona...   
4  2024-11-27 18:58:22  Proactive Gradient Conflict Mitigation in Mult...   

                                            abstract               categories  \
0    The applications of LLM Agents are becoming ...        cs.CL,cs.AI,cs.PL   
1    The recent advancements in auto-regressive m...        cs.AI,cs.CL,cs.CV   
2    Text-to-image diffusion models produce impre...  cs.CV,cs.AI,cs.GR,cs.LG   
3    Coronary artery disease (CAD), one of the le...      eess.IV,cs.AI,cs.CV   
4    Advancing towards generalist agents necessit...        cs.LG,cs.AI,cs.CV   

                                             comment  
0                                      Not available  
1                                      Not available  
2      Project page: https://primecai.github.io/dsd/  
3  Developed by Center for Applied Artificial Int...  
4                                      Not available

### Có bao nhiêu dòng và cột của bộ dữ liệu thô này?

**Kiểm tra bộ dữ liệu đã đạt yêu cầu phải có ít nhất 5 cột và 1000 dòng hay chưa, nếu chưa thì ta sẽ cào bộ dữ liệu mới liên quan đến chủ đề này.**

In [3]:
#ta sẽ tính số dòng và số cột của DataFrame 'raw_df' và lưu trữ nó vào biến 'shape'.
shape = raw_df.shape

In [4]:
print(f"Kích thước dữ liệu hiện tại: {shape}")

if shape[0] >= 1000 and shape[1] >= 5:
    print("Dữ liệu của chúng ta đạt yêu cầu: đủ số dòng và số cột!")
else:
    print("Dữ liệu không đủ. Vui lòng chọn một tập dữ liệu khác!")

Kích thước dữ liệu hiện tại: (257881, 8)
Dữ liệu của chúng ta đạt yêu cầu: đủ số dòng và số cột!


### Mỗi dòng và cột có ý nghĩa gì?

 Mỗi dòng trong tập dữ liệu tương ứng với thông tin về một bài báo khoa học được công bố trên arXiv, bao gồm các cột:

- paper_id: Định danh bài báo, thường là một URL hoặc mã duy nhất của bài báo trên arXiv.
- authors: Danh sách tác giả của bài báo, mỗi tác giả được ngăn cách bởi dấu phẩy.
- updated: Thời gian cập nhật cuối cùng của bài báo, định dạng YYYY-MM-DD HH:MM:SS.
- published: Thời gian xuất bản lần đầu của bài báo, định dạng YYYY-MM-DD HH:MM:SS.
- title: Tiêu đề bài báo, cung cấp cái nhìn tổng quan về nội dung bài báo.
- abstract: Tóm tắt bài báo, cung cấp một đoạn ngắn mô tả nội dung chính.
- categories: Các danh mục mà bài báo thuộc về, giúp phân loại lĩnh vực nghiên cứu.
- comment: Ghi chú bổ sung hoặc thông tin thêm về bài báo (như là bài báo đã được đăng công khai trước đó ở đâu).

#### Dữ liệu thô có dòng nào bị trùng không?

In [5]:
#chúng ta tính toán số dòng bị trùng và lưu trữ nó vào biến 'num_duplicated_rows'.
num_duplicated_rows = raw_df[raw_df.duplicated()].shape[0]

In [6]:
if num_duplicated_rows == 0:
    print("Dữ liệu gốc của bạn không có dòng trùng lặp!")
else:
    if num_duplicated_rows > 1:
        ext = "lines"
    else:
        ext = "line"
    print(f"Dữ liệu gốc của bạn có {num_duplicated_rows} {ext} bị trùng lặp. Vui lòng loại bỏ các dòng trùng lặp!")

Dữ liệu gốc của bạn không có dòng trùng lặp!


In [7]:
raw_df

paper_id  \
0       http://arxiv.org/abs/2411.15100v2   
1       http://arxiv.org/abs/2411.18620v1   
2       http://arxiv.org/abs/2411.18616v1   
3       http://arxiv.org/abs/2411.07976v5   
4       http://arxiv.org/abs/2411.18615v1   
...                                   ...   
257876  http://arxiv.org/abs/2305.13538v1   
257877  http://arxiv.org/abs/2305.13463v1   
257878  http://arxiv.org/abs/2211.01527v2   
257879  http://arxiv.org/abs/2305.13215v1   
257880  http://arxiv.org/abs/2305.12999v1   

                                                  authors  \
0       Yixin Dong,Charlie F. Ruan,Yaxing Cai,Ruihang ...   
1       Zhi Zhang,Srishti Yadav,Fengze Han,Ekaterina S...   
2       Shengqu Cai,Eric Chan,Yunzhi Zhang,Leonidas Gu...   
3       Mahmut S. Gokmen,Caner Ozcan,Moneera N. Haque,...   
4       Zhi Zhang,Jiayi Shen,Congfeng Cao,Gaole Dai,Sh...   
...                                                   ...   
257876                Linwei Sang,Yinliang Xu,Hongbin Sun   
257877                                Taeju Lee,Minkyu Je   
257878  J. Brian Burns,Aravind Sundaresan,Pedro Sequei...   
257879                     Kamal Basulaiman,Masoud Barati   
257880  Savvas Papaioannou,Panayiotis Kolios,Theochari...   

                    updated            published  \
0       2024-11-27 18:59:28  2024-11-22 18:01:37   
1       2024-11-27 18:59:26  2024-11-27 18:59:26   
2       2024-11-27 18:58:52  2024-11-27 18:58:52   
3       2024-11-27 18:58:41  2024-11-12 17:55:39   
4       2024-11-27 18:58:22  2024-11-27 18:58:22   
...                     ...                  ...   
257876  2023-05-22 23:22:13  2023-05-22 23:22:13   
257877  2023-05-22 20:10:56  2023-05-22 20:10:56   
257878  2023-05-22 19:53:33  2022-11-03 00:03:14   
257879  2023-05-22 16:46:37  2023-05-22 16:46:37   
257880  2023-05-22 13:04:34  2023-05-22 13:04:34   

                                                    title  \
0       XGrammar: Flexible and Efficient Structured Ge...   
1       Cross-modal Information Flow in Multimodal Lar...   
2       Diffusion Self-Distillation for Zero-Shot Cust...   
3       DINO-LG: A Task-Specific DINO Model for Corona...   
4       Proactive Gradient Conflict Mitigation in Mult...   
...                                                   ...   
257876  Encoding Carbon Emission Flow in Energy Manage...   
257877  Trend Investigation of Biopotential Recording ...   
257878  Sensor Control for Information Gain in Dynamic...   
257879  Sequence-to-Sequence Forecasting-aided State E...   
257880  Integrated Guidance and Gimbal Control for Cov...   

                                                 abstract  \
0         The applications of LLM Agents are becoming ...   
1         The recent advancements in auto-regressive m...   
2         Text-to-image diffusion models produce impre...   
3         Coronary artery disease (CAD), one of the le...   
4         Advancing towards generalist agents necessit...   
...                                                   ...   
257876    Decarbonizing the energy supply is essential...   
257877    This paper presents the trend of biopotentia...   
257878    We present an approach for autonomous sensor...   
257879    Power system state forecasting has gained mo...   
257880    Coverage path planning with unmanned aerial ...   

                                           categories  \
0                                   cs.CL,cs.AI,cs.PL   
1                                   cs.AI,cs.CL,cs.CV   
2                             cs.CV,cs.AI,cs.GR,cs.LG   
3                                 eess.IV,cs.AI,cs.CV   
4                                   cs.LG,cs.AI,cs.CV   
...                                               ...   
257876                                  eess.SY,cs.SY   
257877                          eess.SY,cs.SY,eess.SP   
257878  cs.AI,cs.CV,cs.SY,eess.SY,I.2.8; I.2.6; I.5.4   
257879                            eess.SY,cs.LG,cs.SY   
257880                                  e

#### Kiểm tra số lượng dòng thiếu dữ liệu trong toàn bộ DataFrame

In [8]:
missing_rows = raw_df.isnull().any(axis=1).sum()

if missing_rows > 0:
    print(f"Có {missing_rows} dòng bị thiếu dữ liệu ở ít nhất một cột. Vui lòng xử lý!")
else:
    print("Dữ liệu hoàn chỉnh, không có dòng nào bị thiếu dữ liệu.")

Có 13 dòng bị thiếu dữ liệu ở ít nhất một cột. Vui lòng xử lý!


#### In ra dòng bị thiếu dữ liệu

In [11]:
missing_data = raw_df[raw_df.isnull().any(axis=1)]
print(missing_data)

                                 paper_id  \
4257    http://arxiv.org/abs/2411.00826v1   
48670   http://arxiv.org/abs/2408.16885v1   
62730   http://arxiv.org/abs/2409.08170v1   
71032   http://arxiv.org/abs/2201.03052v1   
136687  http://arxiv.org/abs/2308.15509v1   
143676  http://arxiv.org/abs/2304.08597v1   
152330  http://arxiv.org/abs/2309.08441v1   
184163  http://arxiv.org/abs/2407.16091v1   
206593  http://arxiv.org/abs/2010.13852v1   
223223  http://arxiv.org/abs/2406.00009v1   
242425  http://arxiv.org/abs/2109.11896v2   
242696  http://arxiv.org/abs/2109.12442v1   
250082  http://arxiv.org/abs/2006.01381v1   

                                                  authors  \
4257    an Zhang,Ming Li,Chun Li,Zhaoxia Liu,Ye Zhang,...   
48670   Ashok Kumar Peepliwall,Hari Mohan Pandey,Surya...   
62730   Santanu Roy,Archit Gupta,Shubhi Tiwari,Palak Sahu   
71032                Richard Apaua,Harjinder Singh Lallie   
136687  Malik Muhammad Qirtas,Evi Zafeiridi,Eleanor Ba...   
143

Ở đây ta có thể thấy cột comment của 13 dòng này không có thông tin và đang là NaN, có thể xử lý bằng cách xóa 13 dòng này hoặc điền và 13 dòng này ở cột commet là 'không có thông tin' 

In [12]:
# điền vào commet của 13 cột 
raw_df['comment'] = raw_df['comment'].fillna('Not available')

Kiểm tra xem còn thiếu dữ liệu hay không?

In [13]:
missing_rows = raw_df.isnull().any(axis=1).sum()

if missing_rows > 0:
    print(f"Có {missing_rows} dòng bị thiếu dữ liệu ở ít nhất một cột. Vui lòng xử lý!")
else:
    print("Dữ liệu hoàn chỉnh, không có dòng nào bị thiếu dữ liệu.")

Dữ liệu hoàn chỉnh, không có dòng nào bị thiếu dữ liệu.


#### Kiểu dữ liệu của mỗi cột hiện tại là gì? Có bao nhiêu cột có kiểu dữ liệu không phù hợp cho việc phân tích sau này?

In [14]:
# chúng ta sẽ xem xét kiểu dữ liệu (dtype) của mỗi cột trong DataFrame 'raw_df' 
# và lưu kết quả vào Series 'dtypes' (Series này có index là tên cột trong DataFrame).
dtypes = raw_df.dtypes

In [15]:
dtypes

paper_id      object
authors       object
updated       object
published     object
title         object
abstract      object
categories    object
comment       object
dtype: object

Hiện tại, chúng ta thấy rằng cột update và published đang ở dạng chuỗi (object). Đây là cột thể hiện ngày tháng, nên chuyển sang kiểu dữ liệu datetime sẽ giúp:

Dễ dàng thực hiện các phép tính liên quan đến thời gian (ví dụ: so sánh ngày, tính khoảng cách giữa các ngày).
Hỗ trợ trực quan hóa dữ liệu dạng chuỗi thời gian (time-series).

Các cột còn lại vẫn giữ nguyên object.

In [16]:
raw_df['updated'] = pd.to_datetime(raw_df['updated'])
raw_df['published'] = pd.to_datetime(raw_df['published'])

In [17]:
dtypes = raw_df.dtypes
dtypes

paper_id              object
authors               object
updated       datetime64[ns]
published     datetime64[ns]
title                 object
abstract              object
categories            object
comment               object
dtype: object

### Đối với mỗi cột có kiểu dữ liệu non-numeric, những giá trị trong các cột đó phân bố như thế nào?

In [18]:
# xem phân bố của các giá trị object.
non_numeric_columns = raw_df.select_dtypes(include='object').columns
unique_values = raw_df[non_numeric_columns].nunique()
print("Số lượng giá trị duy nhất trong mỗi cột non-numeric:")
print(unique_values)

# Tính các thống kê cho cột 'update'
count_u = raw_df['updated'].notnull().sum()  # Số lượng giá trị không thiếu
unique_u = raw_df['updated'].nunique()  # Số lượng giá trị duy nhất
top_u = raw_df['updated'].mode()[0]  # Giá trị xuất hiện nhiều nhất
freq_u = raw_df['updated'].value_counts().max()  # Số lần giá trị xuất hiện nhiều nhất
min_value_u = raw_df['updated'].min()  # Giá trị nhỏ nhất
max_value_u = raw_df['updated'].max()  # Giá trị lớn nhất
mean_u = raw_df['updated'].mean()  # Trung bình
std_u = raw_df['updated'].std()  # Độ lệch chuẩn
percentiles_u = raw_df['updated'].quantile([0.25, 0.5, 0.75])  # Các phân vị

# Tính các thống kê cho cột 'published'
count_p = raw_df['published'].notnull().sum()  # Số lượng giá trị không thiếu
unique_p = raw_df['published'].nunique()  # Số lượng giá trị duy nhất
top_p = raw_df['published'].mode()[0]  # Giá trị xuất hiện nhiều nhất
freq_p = raw_df['published'].value_counts().max()  # Số lần giá trị xuất hiện nhiều nhất
min_value_p = raw_df['published'].min()  # Giá trị nhỏ nhất
max_value_p = raw_df['published'].max()  # Giá trị lớn nhất
mean_p = raw_df['published'].mean()  # Trung bình
std_p = raw_df['published'].std()  # Độ lệch chuẩn
percentiles_p = raw_df['published'].quantile([0.25, 0.5, 0.75])  # Các phân vị

# In ra kết quả
print(f"Count_u: {count_u}, Count_p: {count_p}")
print(f"Unique_u: {unique_u}, unique_p: {unique_p} ")
print(f"Top_u: {top_u}, Top_p: {top_p}")
print(f"Freq_u: {freq_u}, Freq_p: {freq_p}")
print(f"Min_u: {min_value_u}, Min_p: {min_value_p}")
print(f"Max_u: {max_value_u}, Max_p: {max_value_p}")
print(f"Mean_u: {mean_u}, Mean_p: {mean_p}")
print(f"Std_u: {std_u}, Std_p:{std_p}")
print(f"25%_u: {percentiles_u[0.25]}, 25%_p: {percentiles_p[0.25]}")
print(f"50%_u: {percentiles_u[0.5]}, 50%_p: {percentiles_p[0.5]}")
print(f"75%_u: {percentiles_u[0.75]}, 75%_p: {percentiles_p[0.75]}")

Số lượng giá trị duy nhất trong mỗi cột non-numeric:
paper_id      257881
authors       230379
title         257753
abstract      257788
categories     49947
comment       100608
dtype: int64
Count_u: 257881, Count_p: 257881
Unique_u: 257550, unique_p: 257585 
Top_u: 2012-06-27 19:59:59, Top_p: 2012-06-27 19:59:59
Freq_u: 6, Freq_p: 6
Min_u: 1990-04-01 00:00:00, Min_p: 1990-04-01 00:00:00
Max_u: 2024-11-27 18:59:59, Max_p: 2024-11-27 18:59:59
Mean_u: 2021-11-05 03:47:50.417460224, Mean_p: 2021-07-28 01:38:12.083789824
Std_u: 1387 days 08:37:05.054382448, Std_p:1409 days 19:40:57.027689504
25%_u: 2020-11-16 15:52:15, 25%_p: 2020-06-27 02:28:44
50%_u: 2023-04-20 01:46:19, 50%_p: 2022-12-18 05:32:19
75%_u: 2024-05-29 10:23:52, 75%_p: 2024-03-18 11:07:58


Kết quả này cho thấy:

**Số lượng giá trị duy nhất trong các cột non-numeric**:
Các cột như paper_id, authors, title, và abstract có số lượng giá trị duy nhất rất cao (xấp xỉ gần với tổng số dòng). Điều này cho thấy:

- paper_id: Mỗi bài báo có một mã định danh duy nhất.
- authors: Tên tác giả có sự đa dạng rất lớn.
- title: Tiêu đề bài báo hầu hết không trùng lặp.
- abstract: Mô tả ngắn gọn bài báo cũng rất khác biệt.
Cột categories và comment có số lượng giá trị duy nhất ít hơn:

- categories: Có thể chỉ định phân loại theo một số nhóm nhất định (49947 loại khác nhau).
- comment: Một số bài báo không có bình luận hoặc bình luận bị trùng lặp.

**Các chỉ số thời gian liên quan đến hai giá trị (updated và published)**: 

Số lượng và giá trị duy nhất:

- Count_u và Count_p (257881): Tổng số bản ghi trong tập dữ liệu.
- Unique_u (257550) và unique_p (257585): Hầu hết các bản ghi đều có thời gian riêng biệt, chỉ một số ít bị trùng lặp.

**Phân bố thời gian**:
Khoảng thời gian:
- Min_u và Max_u: Dữ liệu trải dài từ năm 1990 đến 2024, cho thấy đây là dữ liệu lâu dài.
- Mean_u và Mean_p: Trung bình ngày nằm trong khoảng 2021, tập trung vào các nghiên cứu gần đây.
- Std_u (1387 ngày) và Std_p (1409 ngày): Thời gian phân phối khá đồng đều trong khoảng thời gian lớn (~4 năm).
Phân vị:
- 25%_u và 25%_p: 25% dữ liệu cũ hơn ngày 2020, còn lại là dữ liệu sau đó.
- 50%_u và 50%_p (median): Giá trị trung vị của dữ liệu rơi vào 2022-2023, cho thấy dữ liệu khá hiện đại.
- 75%_u và 75%_p: Phần lớn dữ liệu thuộc giai đoạn 2024.

### Dữ liệu được thu thập có hợp lý chưa?

In [19]:
raw_df

paper_id  \
0       http://arxiv.org/abs/2411.15100v2   
1       http://arxiv.org/abs/2411.18620v1   
2       http://arxiv.org/abs/2411.18616v1   
3       http://arxiv.org/abs/2411.07976v5   
4       http://arxiv.org/abs/2411.18615v1   
...                                   ...   
257876  http://arxiv.org/abs/2305.13538v1   
257877  http://arxiv.org/abs/2305.13463v1   
257878  http://arxiv.org/abs/2211.01527v2   
257879  http://arxiv.org/abs/2305.13215v1   
257880  http://arxiv.org/abs/2305.12999v1   

                                                  authors             updated  \
0       Yixin Dong,Charlie F. Ruan,Yaxing Cai,Ruihang ... 2024-11-27 18:59:28   
1       Zhi Zhang,Srishti Yadav,Fengze Han,Ekaterina S... 2024-11-27 18:59:26   
2       Shengqu Cai,Eric Chan,Yunzhi Zhang,Leonidas Gu... 2024-11-27 18:58:52   
3       Mahmut S. Gokmen,Caner Ozcan,Moneera N. Haque,... 2024-11-27 18:58:41   
4       Zhi Zhang,Jiayi Shen,Congfeng Cao,Gaole Dai,Sh... 2024-11-27 18:58:22   
...                                                   ...                 ...   
257876                Linwei Sang,Yinliang Xu,Hongbin Sun 2023-05-22 23:22:13   
257877                                Taeju Lee,Minkyu Je 2023-05-22 20:10:56   
257878  J. Brian Burns,Aravind Sundaresan,Pedro Sequei... 2023-05-22 19:53:33   
257879                     Kamal Basulaiman,Masoud Barati 2023-05-22 16:46:37   
257880  Savvas Papaioannou,Panayiotis Kolios,Theochari... 2023-05-22 13:04:34   

                 published                                              title  \
0      2024-11-22 18:01:37  XGrammar: Flexible and Efficient Structured Ge...   
1      2024-11-27 18:59:26  Cross-modal Information Flow in Multimodal Lar...   
2      2024-11-27 18:58:52  Diffusion Self-Distillation for Zero-Shot Cust...   
3      2024-11-12 17:55:39  DINO-LG: A Task-Specific DINO Model for Corona...   
4      2024-11-27 18:58:22  Proactive Gradient Conflict Mitigation in Mult...   
...                    ...                                                ...   
257876 2023-05-22 23:22:13  Encoding Carbon Emission Flow in Energy Manage...   
257877 2023-05-22 20:10:56  Trend Investigation of Biopotential Recording ...   
257878 2022-11-03 00:03:14  Sensor Control for Information Gain in Dynamic...   
257879 2023-05-22 16:46:37  Sequence-to-Sequence Forecasting-aided State E...   
257880 2023-05-22 13:04:34  Integrated Guidance and Gimbal Control for Cov...   

                                                 abstract  \
0         The applications of LLM Agents are becoming ...   
1         The recent advancements in auto-regressive m...   
2         Text-to-image diffusion models produce impre...   
3         Coronary artery disease (CAD), one of the le...   
4         Advancing towards generalist agents necessit...   
...                                                   ...   
257876    Decarbonizing the energy supply is essential...   
257877    This paper presents the trend of biopotentia...   
257878    We present an approach for autonomous sensor...   
257879    Power system state forecasting has gained mo...   
257880    Coverage path planning with unmanned aerial ...   

                                           categories  \
0                                   cs.CL,cs.AI,cs.PL   
1                                   cs.AI,cs.CL,cs.CV   
2                             cs.CV,cs.AI,cs.GR,cs.LG   
3                                 eess.IV,cs.AI,cs.CV   
4                                   cs.LG,cs.AI,cs.CV   
...                                               ...   
257876                                  eess.SY,cs.SY   
257877                          eess.SY,cs.SY,eess.SP   
257878  cs.AI,cs.CV,cs.SY,eess.SY,I.2.8; I.2.6; I.5.4   
257879                            eess.SY,cs.LG,cs.SY   
257880                                  eess.SY,cs.SY   

                                                  comment  
0                                           Not available  
1       

#### Ở đây ta thấy ở cột paper_id là các đường dẫn URL ta có thể lọc lấy định danh duy nhất từ đường dẫn.

In [20]:
raw_df["paper_id"] = raw_df["paper_id"].str.split("/").str[-1]
raw_df

paper_id                                            authors  \
0       2411.15100v2  Yixin Dong,Charlie F. Ruan,Yaxing Cai,Ruihang ...   
1       2411.18620v1  Zhi Zhang,Srishti Yadav,Fengze Han,Ekaterina S...   
2       2411.18616v1  Shengqu Cai,Eric Chan,Yunzhi Zhang,Leonidas Gu...   
3       2411.07976v5  Mahmut S. Gokmen,Caner Ozcan,Moneera N. Haque,...   
4       2411.18615v1  Zhi Zhang,Jiayi Shen,Congfeng Cao,Gaole Dai,Sh...   
...              ...                                                ...   
257876  2305.13538v1                Linwei Sang,Yinliang Xu,Hongbin Sun   
257877  2305.13463v1                                Taeju Lee,Minkyu Je   
257878  2211.01527v2  J. Brian Burns,Aravind Sundaresan,Pedro Sequei...   
257879  2305.13215v1                     Kamal Basulaiman,Masoud Barati   
257880  2305.12999v1  Savvas Papaioannou,Panayiotis Kolios,Theochari...   

                   updated           published  \
0      2024-11-27 18:59:28 2024-11-22 18:01:37   
1      2024-11-27 18:59:26 2024-11-27 18:59:26   
2      2024-11-27 18:58:52 2024-11-27 18:58:52   
3      2024-11-27 18:58:41 2024-11-12 17:55:39   
4      2024-11-27 18:58:22 2024-11-27 18:58:22   
...                    ...                 ...   
257876 2023-05-22 23:22:13 2023-05-22 23:22:13   
257877 2023-05-22 20:10:56 2023-05-22 20:10:56   
257878 2023-05-22 19:53:33 2022-11-03 00:03:14   
257879 2023-05-22 16:46:37 2023-05-22 16:46:37   
257880 2023-05-22 13:04:34 2023-05-22 13:04:34   

                                                    title  \
0       XGrammar: Flexible and Efficient Structured Ge...   
1       Cross-modal Information Flow in Multimodal Lar...   
2       Diffusion Self-Distillation for Zero-Shot Cust...   
3       DINO-LG: A Task-Specific DINO Model for Corona...   
4       Proactive Gradient Conflict Mitigation in Mult...   
...                                                   ...   
257876  Encoding Carbon Emission Flow in Energy Manage...   
257877  Trend Investigation of Biopotential Recording ...   
257878  Sensor Control for Information Gain in Dynamic...   
257879  Sequence-to-Sequence Forecasting-aided State E...   
257880  Integrated Guidance and Gimbal Control for Cov...   

                                                 abstract  \
0         The applications of LLM Agents are becoming ...   
1         The recent advancements in auto-regressive m...   
2         Text-to-image diffusion models produce impre...   
3         Coronary artery disease (CAD), one of the le...   
4         Advancing towards generalist agents necessit...   
...                                                   ...   
257876    Decarbonizing the energy supply is essential...   
257877    This paper presents the trend of biopotentia...   
257878    We present an approach for autonomous sensor...   
257879    Power system state forecasting has gained mo...   
257880    Coverage path planning with unmanned aerial ...   

                                           categories  \
0                                   cs.CL,cs.AI,cs.PL   
1                                   cs.AI,cs.CL,cs.CV   
2                             cs.CV,cs.AI,cs.GR,cs.LG   
3                                 eess.IV,cs.AI,cs.CV   
4                                   cs.LG,cs.AI,cs.CV   
...                                               ...   
257876                                  eess.SY,cs.SY   
257877                          eess.SY,cs.SY,eess.SP   
257878  cs.AI,cs.CV,cs.SY,eess.SY,I.2.8; I.2.6; I.5.4   
257879                            eess.SY,cs.LG,cs.SY   
257880                                  eess.SY,cs.SY   

                                                  comment  
0                                           Not available  
1                                           Not available  
2           Project page: https://primecai.github.io/dsd/  
3       Developed by Center for Applied Artificial Int...  
4                                           Not available  

Đến đây bộ dữ liệu có vẻ đã tốt, ta lưu lại bộ dữ liệu đã được làm sạch này vào file './data/cleaned_Tesla_Stock.csv'.

In [21]:
save_file_name = 'cleaned_arxiv_cs_papers_all.csv'
raw_df.to_csv(save_file_name, index = False)

#### Đọc dữ liệu đã làm sạch ở phần Preprocessing.

In [22]:
file_name = 'cleaned_arxiv_cs_papers_all.csv'
cleaned_df = pd.read_csv(file_name, parse_dates=['updated', 'published'])